### ISSUE TITLE
Provide a mechanism to require a certain Flutter version to build

### ISSUE THREAD

**Issue by rock3r** [flutter/flutter#14229]
Right now there is no way to specify the required Flutter version in the `pubspec.yaml`. One should be able to pin the build to a specific version (or commit) of Flutter so that the build is reproducible in the future. Furthermore, there is no way currently to ensure developers working on a shared codebase all use the same version of Flutter, making collaboration on bugfixing and debugging potentially impossible.

tl;dr it would be nice if it were possible to specify the version of Flutter to use just like one could specify the Dart SDK version to use.

**Comment by rock3r** [comment]
CC @devoncarew

**Comment by Hixie** [comment]
The current suggestion is to use git submodules to check out the specific version of flutter into the app's repository.

**Comment by rock3r** [comment]
Yes I'm aware of that, and there's some [tools](https://github.com/flutter/flutter/issues/12827) to make it less painful to deal with — submodules aren't exactly user friendly to deal with, in the first place, so that is not a great development experience. Then there's the fact that nowhere in the project file it says what version of flutter the code is coded against. It is a problem because you cannot really rely on any non-code documentation (we all know it gets out of date or irrelevant or misleading quite fast).

**Comment by aegis123** [comment]
I also believe this should part of the tooling and the project itself and not some external 3rd party like git. Since maybe I want to use mercurial or svn as cvs system and as developer I should have to freedom to do that without having to be forced to use git for instance.

**Comment by Hixie** [comment]
I think you actually can do this using a flutter SDK constraint:

```yaml
environment:
  sdk: ">=2.0.0-dev.28.0 <3.0.0"
  flutter: ">=0.1.4 <2.0.0"
```

For example, see https://github.com/flutter/plugins/blob/master/packages/image_picker/pubspec.yaml

**Comment by Hixie** [comment]
We should probably document this better.

**Comment by Hixie** [comment]
And test it.

**Comment by bkonyi** [comment]
This is being discussed elsewhere by the Pub team.
